contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -308,17 +308,49 @@ response reduces the complexity of these clients.
 
 ##### Typical status properties
 
-**Conditions** represent the latest available observations of an object's
-state.  They are an extension mechanism intended to be used when the details of
-an observation are not a priori known or would not apply to all instances of a
-given Kind.  For observations that are well known and apply to all instances, a
-regular field is preferred.  An example of a Condition that probably should
-have been a regular field is Pod's "Ready" condition - it is managed by core
-controllers, it is well understood, and it applies to all Pods.
+**Conditions** provide a standard mechanism for higher-level status reporting
+from a controller. They are an extension mechanism which allows tools and other
+controllers to collect summary information about resources without needing to
+understand resource-specific status details. Conditions should complement more
+detailed information about the observed status of an object written by a
+controller, rather than replace it. For example, the "Available" condition of a
+Deployment can be determined by examining `readyReplicas`, `replicas`, and
+other properties of the Deployment. However, the "Available" condition allows
+other components to avoid duplicating the availability logic in the Deployment
+controller.
 
 Objects may report multiple conditions, and new types of conditions may be
 added in the future or by 3rd party controllers. Therefore, conditions are
-represented using a list/slice, where all have similar structure.
+represented using a list/slice of objects, where each condition has a similar
+structure. Conditions are most useful when they follow some consistent
+conventions:
+
+* Conditions should be added to explicitly convey properties that users and
```

Comment:
I don't really agree. I think there's room in the world for two kinds of conditions:
* "Reportables"-- conditions that may or may not be true of particular subsets of items. There's many ways of reporting these and we don't expect consumers to understand them.
* "Consumables"-- these conditions are summaries of the other facts about the object, packaged for consumers that need a high-level bit to trigger off of.

This assumes some sort of bridge controller which understands all the reportables and can summarize them into the consumables.

Reply:
If we allow Condition types to be namespaced (Tim's idea), it would be possible to summarize all namespaced conditions with a consistent polarity using a "bridge controller".

Existing Conditions would need some sort of annotation to indicate their expected polarity, since they are generally mixed.

Reply:
I think there's two ideas here?
* namespacing is usually just to avoid naming collisions. I'm pretty sure we decided against namespacing these way back when specifically to force people to settle on Schelling point names & meanings. I'm not sure if this worked or not. It's definitely going to be confusing to have different conditions with the same name but different namespaces.
* "aspect of": I think many conditions have this relation, where e.g. network reachability is an "aspect of" a pod's overall readiness. If we codified this explicitly, it'd be easy to have a controller aggregate all aspects of a given top-level consumable. This wouldn't necessarily require consistent polarity with the top level consumable (although that'd be most obvious), you could represent with with a signed value, where e.g. -1/0/+1 means false-disables / no effect / true-disables. This could get arbitrarily complicated, though, and I think it might be better to just express it in code in a controller.

Reply:
I'm lost on this thread - I don't know what a "bridge controller" is for?

Regarding polarity, we could have an indicator, as Daniel suggests, so it is self-describing.  But "good/bad" is a pretty vague proclamation.  And we'd have to make it optional ANYWAY, so ...

If the goal is to have machines understand arbitrary conditions, we need an indicator of normal or abnormal polarity.  Is that the goal?  Or do we expect that most consumers of conditions are either a) humans who can read the name and infer polarity; b) programs coded by humans to look for specific conditions with a priori known polarity?

We can't simply say that Conditions are "summarizations" of other info, because they are ALSO the only extension point for external status.

Reply:
I'd like to reach the point where machines can manage arbitrary Conditions. Evolving that from the current state may require additional indicator fields, depending on implementation strategy. If there are existing status contributors with mixed polarity within an object, we'll definitely need a "Polarity" field. If we perform the status aggregation in the "main" controller for the object, we can do this object-by-object and might be able to avoid Polarity.

I interpreted @lavalamp 's idea of a "bridge controller" as a single controller which could operate across all Kinds and perform this aggregation based on mechanical rules. Doing this aggregation within the object's status avoids duplicating the aggregation logic across many clients; doing this in a "bridge controller" avoids duplicating the aggregation in many controllers (which may or may not be as much of an issue).

I'm not sure how many aggregated conditions are needed (vs creating additional objects for "consumable" measurements that are composite). We got away with one in Knative, but singletons always come back to greet you. The advantage of codifying the condition aggregation of "aspect of" in a controller is that it allows new controllers to contribute observations without needing to modify the logic of the main controller to recognize the new conditions.

Playing with an experiment we did in Knative, it would be possible to add "Severity" and "Polarity" fields:

| Severity | Polarity | Effects |
|---|---|---|
| Error (default / empty-string) | Positive | If "False", "Ready" is False |
| Error (default / empty-string) | Negative | If "True", "Ready is False |
| Warning | Positive | No effect on "Ready", but may be highlighted by UIs |
| Warning | Negative | No effect on "Ready", but may be highlighted by UIs, e.g. "ReplicaFailure" in Deployment |
| Info | N/A | Informational observations, such as "Resizing" in PVC |
| unset | unset | Legacy Condition, not aggregated 🗡️ |

🗡️ - For legacy conditions, we might want to prioritize updating those that should be treated as "Error" to set a polarity, then those that should be prioritized as "Warning".

If we proceed object-at-a-time, the Polarity field might be able to be dropped, but this would probably require a certain amount of explicit handling of legacy conditions.

Reply:
I'm not inherently against more self-describing, extensible status, but  tat's a significant change from Conditions today.

Ready is a condition of its own and we don't have "virtual" conditions.  So clients would need to be taught this new structure, which is obviously backwards-incompatible.  CF the design around Pod `readinessGates` which had a constraint that clients who only know to read `Ready` would still be correct.

Reply:
In other words, if you want something like this, I think it needs to be totally additive (a new concept)